prompting/tools/datasets/arxiv.py

```diff
@@ -0,0 +1,227 @@
+# documentation files (the “Software”), to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
+# and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+# the Software.
+
+# THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+# THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+import re
+import sys
+import fitz
+import random
+import datetime
+from dataclasses import dataclass
+import bittensor as bt
+import arxiv
+from typing import Dict, Union, List, Tuple
+from queue import Queue, Full, Empty
+from functools import lru_cache
+from .base import Dataset
+from ..selector import Selector
+
+
+# Create a queue called CACHED_ARTICLES to store arxiv articles that have been fetched
+CACHED_ARTICLES = Queue(maxsize=300)
+
+client = arxiv.Client()
+
+
+@dataclass
+class ArxivPaper:
+
+    # Define common section headers to search for
+    section_headers = [
+        r"(?i)\babstract\b",
+        r"(?i)\bintroduction\b",
+        r"(?i)\bmethods\b",
+        r"(?i)\bmethodology\b",
+        r"(?i)\bresults\b",
+        r"(?i)\bdiscussion\b",
+        r"(?i)\bconclusion\b",
+        r"(?i)\breferences\b",
+    ]
+
+    def __init__(self, paper, text):
+        self.paper = paper
+        self.text = text
+
+        self.parse(text)
+
+    def parse(self, text):
+        # Find sections using regular expressions
+        sections = {}
+        for header in self.section_headers:
+            match = re.search(header, text)
+            if match:
+                sections[match.group().lower()] = match.start()
+
+        # Sort sections by their position in the text
+        sorted_sections = sorted(sections.items(), key=lambda x: x[1])
+
+        # Extract sections
+        paper_sections = {}
+        for i in range(len(sorted_sections)):
+            section_name = sorted_sections[i][0]
+            section_start = sorted_sections[i][1]
+            section_end = (
+                sorted_sections[i + 1][1] if i + 1 < len(sorted_sections) else len(text)
+            )
+            paper_sections[section_name] = text[section_start:section_end].strip(
+                f" \n{section_name}"
+            )
+
+        self.sections = paper_sections
+
+
+# speed up page loading
+@lru_cache(maxsize=1000)
+def _get_text(paper, seed=None) -> arxiv.Result:
```

Comment:
```suggestion
def _get_text(paper: arxiv.Result, seed=None) -> str | None:
```

Let's use typings whether it's possible